tink/controllers/workflow/controller.go

```diff
@@ -0,0 +1,282 @@
+	client.Client
+	WorkflowClient workflowClient
+	Log            logr.Logger
+	Scheme         *runtime.Scheme
+}
+
+// SetupWithManager configures reconciler with a given manager.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager, workflowChan <-chan event.GenericEvent) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&tinkv1alpha1.Workflow{}).
+		Watches(
+			&source.Channel{Source: workflowChan},
+			&handler.EnqueueRequestForObject{},
+		).
+		Complete(r)
+}
+
+// +kubebuilder:rbac:groups=tinkerbell.org,resources=workflows;workflows/status,verbs=get;list;watch;create;update;patch;delete
+
+// Reconcile ensures state of Tinkerbell workflows.
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	logger := r.Log.WithValues("workflow", req.NamespacedName.Name)
+
+	// Fetch the workflow.
+	workflow := &tinkv1alpha1.Workflow{}
+	if err := r.Get(ctx, req.NamespacedName, workflow); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		logger.Error(err, "Failed to get workflow")
+
+		return ctrl.Result{}, fmt.Errorf("failed to get workflow: %w", err)
+	}
+
+	// Ensure that we add the finalizer to the resource
+	if err := common.EnsureFinalizer(ctx, r.Client, logger, workflow, tinkv1alpha1.WorkflowFinalizer); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to ensure finalizer on workflow: %w", err)
+	}
+
+	// Handle deleted wokflows.
+	if !workflow.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, workflow)
+	}
+
+	return r.reconcileNormal(ctx, workflow)
+}
+
+func (r *Reconciler) reconcileNormal(ctx context.Context, w *tinkv1alpha1.Workflow) (ctrl.Result, error) {
+	logger := r.Log.WithValues("workflow", w.Name)
+
+	workflowID := w.TinkID()
+
+	if workflowID == "" {
+		id, err := r.createWorkflow(ctx, w)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		workflowID = id
+	}
+
+	tinkWorkflow, err := r.WorkflowClient.Get(ctx, workflowID)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get workflow: %w", err)
+	}
+
+	// Make sure that we record the tinkerbell id for the workflow
+	patch := client.MergeFrom(w.DeepCopy())
+	w.SetTinkID(workflowID)
+
+	if err := r.Client.Patch(ctx, w, patch); err != nil {
+		logger.Error(err, "Failed to patch workflow")
+
+		return ctrl.Result{}, fmt.Errorf("failed to patch workflow: %w", err)
+	}
+
+	return r.reconcileStatus(ctx, w, tinkWorkflow)
+}
+
+func (r *Reconciler) reconcileStatus(ctx context.Context, w *tinkv1alpha1.Workflow, tinkWorkflow *workflow.Workflow) (ctrl.Result, error) {
+	logger := r.Log.WithValues("workflow", w.Name)
+	patch := client.MergeFrom(w.DeepCopy())
+
+	// Try to patch what we have even if we hit errors gather additional status information
+	defer func() {
+		if err := r.Client.Status().Patch(ctx, w, patch); err != nil {
+			logger.Error(err, "Failed to patch workflow status")
+		}
+	}()
+
+	w.Status.Data = tinkWorkflow.GetData()
```

Comment:
Hm, I guess this is content of rendered workflow, right? If so, it may contain secrets. And it will contain secrets when we use it for CAPT.

Should we then create a secret for each workflow and store the data there? This actually also applies to Templates and hardware selectors :/